### ISSUE TITLE
Cumsum slower than NumPy on CPU

### ISSUE THREAD

**Issue by dionhaefner** [google/jax#4135]
I noticed that `cumsum` is several times slower than NumPy on 1 CPU, even on large arrays.

`jax.lax.scan` is faster, but only works along the leading axis (#2509).

(this might be connected to #4134)

#### Example script

```python
import jax
jax.config.update('jax_enable_x64', True)

import numpy as np
import jax.numpy as jnp


@jax.jit
def bench_cumsum(arr):
    return jnp.cumsum(arr, axis=0)


@jax.jit
def bench_scan(arr):
    def accumulate(carry, y):
        new_y = y + carry
        return new_y, new_y

    init = jnp.zeros(arr.shape[1:], dtype=arr.dtype)
    return jax.lax.scan(accumulate, init, arr)[1]


def bench_np(arr):
    return np.cumsum(arr, axis=0)


if __name__ == '__main__':
    import time
    shape = (360, 160, 120, 3)

    key = jax.random.PRNGKey(17)
    arr = jax.random.uniform(key, shape=shape, dtype='float64')

    res_cumsum = bench_cumsum(arr).block_until_ready()
    start = time.perf_counter()
    for _ in range(10):
        bench_cumsum(arr).block_until_ready()
    end = time.perf_counter()
    print(f'bench_cumsum: {end-start:.3f}s')

    res_scan = bench_scan(arr).block_until_ready()
    start = time.perf_counter()
    for _ in range(10):
        bench_scan(arr).block_until_ready()
    end = time.perf_counter()
    print(f'bench_scan: {end-start:.3f}s')
    np.testing.assert_allclose(res_cumsum, res_scan)

    arr_np = np.array(arr)
    start = time.perf_counter()
    for _ in range(10):
        bench_np(arr_np)
    end = time.perf_counter()
    print(f'bench_np: {end-start:.3f}s')

```

#### Output

```
$ taskset -c 0 python bench_cumsum.py
bench_cumsum: 6.635s
bench_scan: 0.885s
bench_np: 2.230s
```

**Comment by mattjj** [comment]
Thanks for the excellent benchmarking script! This certainly sounds like a performance bug.

It's true without 64bit mode too; I got these timings with 32bit (commenting out the 64bit-enable liine):

```
bench_cumsum: 3.366s
bench_scan: 0.183s
bench_np: 0.749s
```

My timings with 64bit were similar to yours (running the original script):

```
bench_cumsum: 7.550s
bench_scan: 0.382s
bench_np: 1.618s
```

**Comment by mattjj** [comment]
I ran with `JAX_LOG_COMPILES=1` and that suggests to me this is not a recompilation issue.

We use a [parallel prefix scan](https://github.com/google/jax/blob/6d54eb563eebd4bfecfe7b6f33aa284e280c5320/jax/lax/lax.py#L5149-L5168) to implement cumulative sum on all backends. I'm wondering if it's slow on CPU for some reason, and maybe the previous reduce-window version would be faster.

**Comment by hawkinsp** [comment]
FWIW, I get the following timings on my workstation with:
* `jax`: 0.2.3
* `jaxlib`: 0.1.56
* `numpy`: 1.19.2

are:

```
$ taskset -c 0 python cs.py
/usr/local/google/home/phawkins/.pyenv/versions/py373a/lib/python3.7/site-packages/jax/lib/xla_bridge.py:130: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
bench_cumsum: 1.989s
bench_scan: 0.436s
bench_np: 1.674s
$ python cs.py
bench_cumsum: 0.895s
bench_scan: 0.362s
bench_np: 1.623s
```

i.e., JAX is a little bit slower when forced to use a single core, but not much, and significantly faster if allowed to use multiple cores.

**Comment by dionhaefner** [comment]
Interesting. But doesn’t the still much faster scan indicate how much performance is left on the table?

**Comment by hawkinsp** [comment]
#4603 improves `jnp.cumsum`. It's possible there's more we can do there. I'm a bit reluctant to use `lax.scan` as the implementation however, because my suspicion is the associative version is more numerically stable for long sequences. Perhaps there is something to using it for short sequences though.

**Comment by dionhaefner** [comment]
Looks very nice. I'll try this ASAP!

**Comment by dionhaefner** [comment]
Very nice indeed, especially the parallel performance. This is what I get on our cluster:

```bash
$ python test.py
bench_cumsum: 0.556s  # 🚀!
bench_scan: 0.697s
bench_np: 2.611s
```

```bash
$ taskset -c 0 python test.py
bench_cumsum: 2.017s
bench_scan: 0.895s
bench_np: 2.412s
```

I personally care more about sequential performance but I guess I can always use scan in this case (as soon as  #4591 is merged).
